ariadne/asgi/handlers/graphql_ws.py

```diff
@@ -0,0 +1,293 @@
+        except Exception as error:
+            log_error(error, self.logger)
+
+            if isinstance(error, WebSocketConnectionError):
+                payload = error.payload  # pylint: disable=no-member
+            else:
+                payload = {"message": "Unexpected error has occurred."}
+
+            await websocket.send_json(
+                {"type": GraphQLWSHandler.GQL_CONNECTION_ERROR, "payload": payload}
+            )
+            await websocket.close()
+
+    async def handle_websocket_connection_terminate_message(self, websocket: WebSocket):
+        await websocket.close()
+
+    async def keep_websocket_alive(self, websocket: WebSocket):
+        if not self.keepalive:
+            return
+        while websocket.application_state != WebSocketState.DISCONNECTED:
+            try:
+                await websocket.send_json(
+                    {"type": GraphQLWSHandler.GQL_CONNECTION_KEEP_ALIVE}
+                )
+            except WebSocketDisconnect:
+                return
+            await asyncio.sleep(self.keepalive)
+
+    async def start_websocket_operation(
+        self,
+        websocket: WebSocket,
+        data: Any,
+        operation_id: str,
+        operations: Dict[str, Operation],
+    ):
+        if self.schema is None:
+            raise TypeError("schema is not set, call configure method to initialize it")
+        context_value = await self.get_context_for_request(websocket)
+
+        success, results = await subscribe(
+            self.schema,
+            data,
+            context_value=context_value,
+            root_value=self.root_value,
+            validation_rules=self.validation_rules,
+            debug=self.debug,
+            introspection=self.introspection,
+            logger=self.logger,
+            error_formatter=self.error_formatter,
+        )
+
+        if not success:
+            results = cast(List[dict], results)
+            await websocket.send_json(
+                {
+                    "type": GraphQLWSHandler.GQL_ERROR,
+                    "id": operation_id,
+                    "payload": results[0],
+                }
+            )
+        else:
+            results = cast(AsyncGenerator, results)
+            operations[operation_id] = Operation(
+                id=operation_id,
+                name=data.get("operationName"),
+                generator=results,
+            )
+
+            if self.on_operation:
+                try:
+                    result = self.on_operation(websocket, operations[operation_id])
+                    if result and isawaitable(result):
+                        await result
+                except Exception as error:
+                    if not isinstance(error, GraphQLError):
+                        error = GraphQLError(str(error), original_error=error)
+                    log_error(error, self.logger)
+
+            asyncio.ensure_future(
```

Comment:
does this code "work"? `observe_async_results` returns `None` and `asyncio.ensure_future` will raise when passed `None` (see: https://docs.python.org/3/library/asyncio-future.html#asyncio.ensure_future).